WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -0,0 +1,191 @@
+		private SecureRandom SecureRandom { get; }
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private Coin[] Coins { get; set; }
+
+		public CoinJoinClient(
+			Round round,
+			IArenaRequestHandler arenaRequestHandler,
+			BitcoinSecret bitcoinSecret,
+			IEnumerable<Coin> coins,
+			KeyManager keymanager)
+		{
+			Round = round;
+			ArenaRequestHandler = arenaRequestHandler;
+			BitcoinSecret = bitcoinSecret;
+			Keymanager = keymanager;
+			SecureRandom = new SecureRandom();
+			Coins = coins.ToArray();
+		}
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			try
+			{
+				// Register coins.
+				AliceClient[] aliceClients = await RegisterCoinsAsync().ConfigureAwait(false);
+
+				// Confirm coins.
+				await ConfirmConnectionsAsync(aliceClients, stoppingToken).ConfigureAwait(false);
+
+				Money[]? othersInputAmounts = null; // TODO: Get it from somewhere.
+
+				// Planning
+				var outputs = await PlanAmountsAsync(othersInputAmounts, stoppingToken).ConfigureAwait(false);
+
+				// Output registration.
+				await RegisterOutputsAsync(outputs).ConfigureAwait(false);
+
+				Transaction? unsignedCoinJoinTransaction = null; // TODO: Get it from somewhere.
+
+				// Send signature.
+				await SignTransactionAsync(aliceClients, unsignedCoinJoinTransaction).ConfigureAwait(false);
+			}
+			catch (Exception ex)
+			{
+			}
+		}
+
+		public async Task StartMixingCoinsAsync()
+		{
+			await StartAsync(DisposeCts.Token).ConfigureAwait(false);
+		}
+
+		private async Task<AliceClient[]> RegisterCoinsAsync()
+		{
+			var aliceArenaClient = new ArenaClient(
+				Round.AmountCredentialIssuerParameters,
+				Round.VsizeCredentialIssuerParameters,
+				AmountCredentialPool,
+				VsizeCredentialPool,
+				ArenaRequestHandler,
+				SecureRandom);
+
+			List<AliceClient> aliceClients = new();
+			try
+			{
+				foreach (var coin in Coins)
+				{
+					// Parallelize or Random delay?
+					aliceClients.Add(await AliceClient.CreateNewAsync(aliceArenaClient, coin, BitcoinSecret, Round.Id, Round.Hash, Round.FeeRate).ConfigureAwait(false));
+				}
+			}
+			catch (Exception)
+			{
+				foreach (var alice in aliceClients)
+				{
+					// Remove already registered Inputs.
+					try
+					{
+						await alice.RemoveInputAsync().ConfigureAwait(false);
+					}
+					catch (Exception)
+					{
+						// Log?
+					}
+				}
+				throw;
+			}
+
+			return aliceClients.ToArray();
+		}
+
+		private async Task ConfirmConnectionsAsync(AliceClient[] aliceClients, CancellationToken stoppingToken)
+		{
+			Task[] confirmTasks = aliceClients.Select(aliceClient => aliceClient.ConfirmConnectionAsync(TimeSpan.FromSeconds(5), stoppingToken)).ToArray();
+
+			await Task.WhenAll(confirmTasks).ConfigureAwait(false);
+
+			var exceptions = confirmTasks.Where(t => t.IsFaulted && t.Exception is { }).Select(t => t.Exception);
+			if (exceptions.Any())
+			{
+				// Error! Try to de-register inputs?
+			}
```

Comment:
This is how I think the error handler could be implemented: https://thesharperdev.com/csharps-whenall-and-exception-handling/ (see the end of the article)

Reply:
I agree, but see also https://github.com/zkSNACKs/WalletWasabi/pull/5694#discussion_r622485196, there isn't much to be done in case of failure.

Reply:
I almost forget. Parallelism is wrong here too. Or `WabiSabiClient` had to be made thread-safe.

Reply:
`WabiSabiClient` is stateless, it is thread-safe.

Reply:
`WabiSabiClient` is stateless, it is thread-safe.